core/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -93,7 +95,27 @@ protected void doExecute(SearchRequest searchRequest, ActionListener<SearchRespo
             logger.debug("failed to optimize search type, continue as normal", e);
         }
 
-        searchAsyncAction(searchRequest, listener).start();
+        ActionListener<SearchResponse> wrapper = listener;
```

Comment:
maybe `final ActionListener<SearchResponse> wrapper;` and assign `listener` in an else block?